falcon/asgi/request.py

```diff
@@ -712,16 +715,38 @@ async def get_media(self):
 
         See also :ref:`media` for more information regarding media handling.
 
+        Note:
+            When ``get_media`` is called on a request with an empty body,
+            Falcon will let the media handler try to deserialize the body
+            and will return the value returned by the handler or propagate
+            the exception raised by it. To instead return a different value
+            in case of an exception by the handler, specify the argument
+            ``when_empty_fallback``.
+
         Warning:
             This operation will consume the request stream the first time
             it's called and cache the results. Follow-up calls will just
             retrieve a cached version of the object.
 
+        Args:
+            when_empty_fallback: Fallback value to return when there is no body
+                in the request and the media handler raises an error
+                (like in the case of the default JSON media handler).
+                By default, Falcon uses the value returned by the media handler
+                or propagates the raised exception, if any.
+
         Returns:
             media (object): The deserialized media representation.
         """
 
-        if self._media is not None or self.stream.eof:
+        if self._media_error is not None:
+            if when_empty_fallback is _UNSET or not isinstance(
+                self._media_error, errors.MediaNotFoundError
+            ):
+                raise self._media_error
+            self._media = when_empty_fallback
+            self._media_error = None
```

Comment:
It seems like we might want to avoid introducing potentially surprising side effects when the empty fallback is provided. As an app developer I would personally expect that if I called the method without passing `when_empty_fallback`, an error would be raised consistently on an empty body, even if somewhere else (say in a middleware component) the method was invoked with a fallback value. My mental model is more like `dict.get()` vs. `dict.setdefault()` (albeit that the analogy is a bit of a stretch).

Reply:
that was done deliberately, but I see your point.

the though was that `when_empty_fallback` takes the place of the data returned by the handler. Instead you are suggesting we need to supply it the every time since we don't cache it. So also here https://github.com/falconry/falcon/blob/6d0d2bd67ac8a34d256e932e24639dfcab544f07/falcon/asgi/request.py#L763-L767 we would not cache it?

Because otherwise it would not consistent the first invocation vrt the subsequent ones

Reply:
I'm with @kgriffs on this one. :arrow_up: If we support this machinery for `Request.get_media()` (as the `when_empty_fallback` _kwarg_), that is.

Reply:
So what I was suggesting was that we would not cache `when_empty_fallback` but just use it if supplied, similar to how the `default` kwarg works for `get_header()` et al. (note also that taking this approach would make `get_media()` more consistent with the rest of the interface).

I.e.:

```py
# This is the first time get_media is accessed. We try to read in the request body
#   and discover it is empty. We remember this fact. But then we see that
#   a value was given for when_empty_fallback, so we return that value.
a = req.get_media(when_empty_fallback=None)
assert a == None

# This time around we find that the body was previously determined to be empty.
#   So we don't try to deserialize it, and just return the fallback instead.
b = req.get_media(when_empty_fallback={})
assert b == {}

# Yet again we find that the body is known to be empty. But this time we do
#   not have a fallback value, so we need to raise an error to the app. The
#   app can then decide what to do about this situation (or just let the default
#   400 error bubble up) since by not providing a fallback value, it is implied 
#   that the client violated the API contract (a body is required).
#
# I would not expect {} to be returned here. Granted, it is an edge case to
#   have an app that is reading from req.get_media() multiple times in different 
#   ways, but when it comes to side effects, it is often prudent to be better 
#   safe than sorry.
with pytest.raises(errors.MediaNotFoundError):
    req.get_media()
```

Reply:
Reposting my thoughts from Gitter where I was basically stating the same, it feels:
> If we supported some sort of fallback on the handler level, then it would make sense that it would be materialized for subsequent invocations without explicit parameters.
> But if it is just a `req.get_media()` customization, it should be altering `get_media()`'s behaviour for that invocation, not inscribing any side effects for the remaining lifetime of the request-response cycle.

Reply:
makes sense